### PR TITLE
Duplicate existing class to add comments.

### PR DESCRIPTION
**Pull request by drn**
I duplicated `sb_sudoku.rb` and renamed it `sudoku.rb` so I could create this PR and leave comments for you!

### CODE REVIEW

File: sudoku.rb [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+  #
+  # @return [Boolean] true if board is valid
+  #
+  def solved?
+    #
+    # TODO
+    #
+
+    board = @board
```
Comment (drn) [inline sudoku.rb:58]: instead of declaring a local variable, you can just leverage the `@board` variable. if you did want a cleaner accessor, check out `attr_reader` which essentially creates an accessor method for `@board`
not much of a difference, but skips a line of code or moves the handling outside of the method, reducing method complexity

File: sudoku.rb [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    row_count = 0
+    valid_count = 0
+    valid = false
+
+    # check if vals in rows are unique
+    uniq_check = Proc.new do |b|
```
Comment (drn) [inline sudoku.rb:64]: procs / lambdas / blocks - great tools when necessary. however, having the block modify a variable outside the context of proc makes the logic a little hard to parse

an alternative here is to declare a private helper method that returns whether or not the input set of values are valid and return true / false

File: sudoku.rb [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+      b.each do |row|
+        row_count += 1
+        valid_count += 1 if row == row.uniq
+      end
+      valid_count == row_count ? valid = true : valid = false
+    end
+
+    # check if vals in columns are unique
+    columns = board.transpose
+    uniq_check.call(columns)
```
Comment (drn) [inline sudoku.rb:74]: we need to verify that rows are checked as well. the example test case would still pass without it though - let's modify it so that the failure example covers the invalid row case

Comment (drn) [reply sudoku.rb:74]: we can short circuit the logic here if this returns false since it wont change the return value even if the box checks returned true

File: sudoku.rb [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+    # check if each 3x3 box is unique
+    # creates new array from each 3x3 box
+    box1 = board[0][0..2].concat(board[1][0..2], board[2][0..2])
```
Comment (drn) [inline sudoku.rb:78]: instead of a explicitly assembling the boxes, any way we can do this programmatically?

File: sudoku.rb (hunk 2 above)
Comment (drn) [reply sudoku.rb:64]: generally single letter variable names are hard to understand - would recommend choosing something more descriptive

File: sudoku.rb [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+  # @param board [Array]
+  #
+  def initialize(board)
+    @board = board
+  end
+
+  #
+  # @return [Boolean] true if board is valid
+  #
+  def solved?
```
Comment (drn) [inline sudoku.rb:53]: when writing production code, one of the main goals to focus on is "how easy is this for another engineer to understand?". it's sometimes a little fuzzy how to do this, but there are a number of best practices / benchmarks to follow

one way of making code easy to understand is breaking a large method out into smaller helper methods. in ruby, the community strives to make the code read as close to english as possible - meaning, the helper methods should verify explicitly describe what it does. for this method, we can break out the check into a method, row checks, column checks, box checks all into separate methods

side note - there are various measures of complexity that developers can help benchmark there code with. [cyclomatic complexity](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Metrics/CyclomaticComplexity), [perceived complexity](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Metrics/PerceivedComplexity), [abc method size](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Metrics/AbcSize), etc

a good benchmark is that a method shouldn't ever be over 10 lines long

generally, we shoot for small classes and methods that serve a single purpose

Comment (drn) [reply sudoku.rb:53]: for example, here is an example snippet from the `Program::Earn` class in the thanx codebase:
```
  def issue_to(user)
    @user = user
    user.redis_lock do
      return nil unless issuable?
      create_reward
      send_email
      send_notification
    end
  end

  def ensure
    retire_rewards
    return unless program.active?
    return unless program.earnable?
    finalize_target
    target.call
    issue!
    finalize_program
  end
```

the majority of the logic are private helper methods or calls to separate business-logic encapsulating classes

File: sudoku.rb [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    #
+
+    board = @board
+    row_count = 0
+    valid_count = 0
+    valid = false
+
+    # check if vals in rows are unique
+    uniq_check = Proc.new do |b|
+      b.each do |row|
```
Comment (drn) [inline sudoku.rb:65]: instead of passing in the collection of arrays to check, another option is to declare a method that checks a single array. that would enable us to break out of the method if any of the rows failed to meet the check + would let us skip the valid_count check (if we want to assert input board sizing, that could even be a separate check instead of coupling the board size check with the array contents check). would potentially save a large number of operations from being run

**Review by drn — COMMENTED** [review]
@izarredondo this is great stuff!!

first, want to thank you for pushing forward on your learning here.
second, keep up the great work and dont be discouraged by lengthy comments. im consciously being nitpicky here to over-communicate some of the concepts and reviewing this as i would any PR that any of the engineers on the team would submit.

for any other changes you'd like me to get feedback on, open up a PR with the given changes and use github's "request review" functionality to tag me :)

**Comment by drn** [comment]
@izarredondo going to close this PR so it doesn't show up in my PR inbox (http://github.com/pulls) - however, I get emails for all activity / comments on github, so feel free to add comments here or open up a new PR next time you're looking for feedback

**Comment by izarredondo** [comment]
@drn Noted, thank you so much for the feedback!
